MediaBrowser.Providers/TV/SeriesMetadataService.cs

```diff
@@ -140,38 +140,30 @@ private void RemoveObsoleteSeasons(Series series)
 
         private void RemoveObsoleteEpisodes(Series series)
         {
-            var episodes = series.GetEpisodes(null, new DtoOptions(), true).OfType<Episode>().ToList();
-            var numberOfEpisodes = episodes.Count;
-            // TODO: O(n^2), but can it be done faster without overcomplicating it?
-            for (var i = 0; i < numberOfEpisodes; i++)
-            {
-                var currentEpisode = episodes[i];
-                // The outer loop only examines virtual episodes
-                if (!currentEpisode.IsVirtualItem)
-                {
-                    continue;
-                }
+            var episodesBySeason = series.GetEpisodes(null, new DtoOptions(), true)
+                            .OfType<Episode>()
+                            .GroupBy(e => e.ParentIndexNumber)
+                            .ToList();
 
-                // Virtual episodes without an episode number are practically orphaned and should be deleted
-                if (!currentEpisode.IndexNumber.HasValue)
-                {
-                    DeleteEpisode(currentEpisode);
-                    continue;
-                }
-
-                for (var j = i + 1; j < numberOfEpisodes; j++)
+            foreach (var seasonEpisodes in episodesBySeason)
+            {
+                // Only consider non-physical episodes
+                var nonPhysicalEpisodes = seasonEpisodes.Where(e => e.IsVirtualItem || e.IsMissingEpisode).ToList();
+                var physicalEpisodes = seasonEpisodes.Except(nonPhysicalEpisodes);
```

Comment:
I feel like there should exist a better way of doing this...
Worst case we write something ourselves.